src/DotVVM.Framework/Compilation/Binding/MemberExpressionFactory.cs

```diff
@@ -0,0 +1,457 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Dynamic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using DotVVM.Framework.Binding;
+using DotVVM.Framework.Controls;
+using DotVVM.Framework.Utils;
+using Microsoft.CSharp.RuntimeBinder;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace DotVVM.Framework.Compilation.Binding
+{
+    public class MemberExpressionFactory
+    {
+        private readonly IExtensionsProvider extensionsProvider;
+
+        public MemberExpressionFactory(IServiceProvider serviceProvider)
+        {
+            extensionsProvider = serviceProvider.GetService<IExtensionsProvider>();
+            if (extensionsProvider == null)
+                extensionsProvider = new DefaultExtensionsProvider();
+        }
+
+        public Expression GetMember(Expression target, string name, Type[] typeArguments = null, bool throwExceptions = true, bool onlyMemberTypes = false)
+        {
+            if (target is MethodGroupExpression)
+                throw new Exception("Can not access member on method group.");
+
+            var type = target.Type;
+            if (type == typeof(UnknownTypeSentinel)) if (throwExceptions) throw new Exception($"Type of '{target}' could not be resolved."); else return null;
+
+            var isStatic = target is StaticClassIdentifierExpression;
+
+            var isGeneric = typeArguments != null && typeArguments.Length != 0;
+            var genericName = isGeneric ? $"{name}`{typeArguments.Length}" : name;
+
+            if (!isGeneric && !onlyMemberTypes && typeof(DotvvmBindableObject).IsAssignableFrom(target.Type) &&
+                GetDotvvmPropertyMember(target, name) is Expression result) return result;
+
+            var members = type.GetAllMembers(BindingFlags.Public | (isStatic ? BindingFlags.Static : BindingFlags.Instance))
+                .Where(m => ((isGeneric && m is TypeInfo) ? genericName : name) == m.Name)
+                .ToArray();
+
+            if (members.Length == 0)
+            {
+                // We did not find any match in regular methods => try extension methods
+                var extensions = extensionsProvider.GetExtensionMethods()
```

Comment:
Shall we cache this? I am afraid that this will browse the types and look for the methods every single time.

Reply:
We probably should. The implementation now caches the result of the method call `GetMethods(..)` to avoid unnecessary reflection. Also the attribute `ExtensionAttribute` is now checked to ensure that we yield only extension methods